app/controllers/application_controller.rb

```diff
@@ -5,6 +5,75 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions unless test?
+    set :session_secret, "wut"
   end
 
-end
\ No newline at end of file
+  get '/' do
+    @users = User.all
+    @tweets = Tweet.all
+    erb :welcome
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :"/users/show"
+  end
+
+  get '/signup' do
+    if !logged_in?
+      erb :signup
+    else
+      redirect '/tweets'
+    end
+  end
+
+  get '/login' do
+    if !logged_in?
+      erb :login
+    else
+      redirect '/tweets'
+    end
+  end
+
+  get '/logout' do
+    if logged_in?
+      session.clear
+      redirect '/login'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/signup' do
+    user = User.new(:username => params[:username],:email => params[:email],
+     :password => params[:password])
+    if user.save
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/signup"
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect '/login'
+    end
+  end
+
+  helpers do
+    def current_user
+      User.find(session[:user_id])
+    end
+
+    def logged_in?
+      !!session[:user_id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```